### PR TITLE
Micrometer Observation Support

### PR DESCRIPTION
**Pull request by n0tl3ss**
- This PR instruments: https://micrometer.io/docs/observation for metrics and traces.

- Here is an example of two applications:  `observation-test` and `observation-test2`. Both have `POST` `/log` endpoint but `observation-test` calls `observation-test2`. Client application has to create two beans: `io.micrometer.tracing.propagation.Propagator` and `io.micrometer.tracing.Tracer` to get same result:
image

### CODE REVIEW

File: micrometer-observation/src/main/java/io/micronaut/micrometer/observation/DefaultObservedFactory.java [hunk 1]
```diff
@@ -0,0 +1,127 @@
...
+ *     <li>{@link ObservationHandler} - Observation Handlers for exporting metrics, traces</li>
+ *     <li>{@link ObservationFilter} - ObservationFilter that adds common low cardinality key values to all observations</li>
+ * </ul>
+ */
+@Factory
+@SuppressWarnings("rawtypes")
+public final class DefaultObservedFactory {
+    @Singleton
+    ObservationRegistry observationRegistry(
+        @Nullable List<ObservationPredicate> observationPredicates,
```
Comment (dstepanov) [inline micrometer-observation/src/main/java/io/micronaut/micrometer/observation/DefaultObservedFactory.java:54]: I don't think we need nullable

File: micrometer-observation/src/main/java/io/micronaut/micrometer/observation/http/server/ObservationServerFilter.java [hunk 2]
```diff
@@ -0,0 +1,107 @@
...
+                            e -> onError(mutableHttpResponse, context, observation, e), () -> {
+                                if (mutableHttpResponse.status().getCode() >= 400) {
+                                    onError(mutableHttpResponse, context, observation, null);
+                                } else {
+                                    context.setResponse(mutableHttpResponse);
+                                    observation.stop();
+                                }
+                            }))
+                    .doOnError(throwable -> onError(null, context, observation, throwable))
+                    .contextWrite(ctx -> ReactorPropagation.addPropagatedContext(ctx, propagatedContext));
```
Comment (dstepanov) [inline micrometer-observation/src/main/java/io/micronaut/micrometer/observation/http/server/ObservationServerFilter.java:96]: `ReactorPropagation.addPropagatedContext` is not needed anymore

**Review by dstepanov — CHANGES_REQUESTED** [review]

File: micrometer-observation/src/main/java/io/micronaut/micrometer/observation/DefaultObservedFactory.java [hunk 3]
```diff
@@ -0,0 +1,113 @@
...
+public final class DefaultObservedFactory {
+    @Singleton
+    ObservationRegistry observationRegistry(
+        List<ObservationPredicate> observationPredicates,
+        List<GlobalObservationConvention<?>> observationConventions,
+        List<ObservationHandler<?>> observationHandlers,
+        List<ObservationFilter> observationFilters
+    ) {
+        ObservationRegistry observationRegistry = ObservationRegistry.create();
+        observationHandlers.forEach(observationRegistry.observationConfig()::observationHandler);
```
Comment (marcingrzejszczak) [inline micrometer-observation/src/main/java/io/micronaut/micrometer/observation/DefaultObservedFactory.java:60]: I don't think this is enough. You should group all `TracingObservationHandlers` together with a `FirstMatchingCompositeHandler`. Same for `MeterObservationHandler`. Check this out https://github.com/spring-projects/spring-boot/blob/main/spring-boot-project/spring-boot-actuator-autoconfigure/src/main/java/org/springframework/boot/actuate/autoconfigure/observation/ObservationAutoConfiguration.java#L85-L107

File: micrometer-observation/src/main/java/io/micronaut/micrometer/observation/DefaultObservedFactory.java [hunk 4]
```diff
@@ -0,0 +1,113 @@
...
+        observationPredicates.forEach(observationRegistry.observationConfig()::observationPredicate);
+        observationFilters.forEach(observationRegistry.observationConfig()::observationFilter);
+        observationConventions.forEach(observationRegistry.observationConfig()::observationConvention);
+        return observationRegistry;
+    }
+
+    @Singleton
+    @Requires(classes = Tracer.class)
+    @Requires(bean = Tracer.class)
+    public ObservationHandler<?> defaultTracingObservationHandler(Tracer tracer) {
```
Comment (marcingrzejszczak) [inline micrometer-observation/src/main/java/io/micronaut/micrometer/observation/DefaultObservedFactory.java:70]: This has to be last in the `TracingObservationHandler` list AFTER `sender` and `receiver`

**Review by marcingrzejszczak — CHANGES_REQUESTED** [review]
Added a remark related to groupings of handlers, other than that LGTM! Great job!

**Comment by dstepanov** [comment]
I would extract the HTTP stuff into a particular module.

**Review by dstepanov — APPROVED** [review]

**Comment by graemerocher** [comment]
this should bump the version to 5.1 snapshot

**Review by marcingrzejszczak — APPROVED** [review]

**Review by graemerocher — APPROVED** [review]

**Comment by marcingrzejszczak** [comment]
:partying_face: congrats to everyone :)

**Comment by jonatan-ivanov** [comment]
Thanks for adding this!
Should this issue be closed? https://github.com/micronaut-projects/micronaut-micrometer/issues/492

**Pull request merged**